test/e2e/cluster_upgrade.go

```
@@ -0,0 +1,122 @@
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Skipped", func() {
+	Describe("Cluster upgrade", func() {
+		name := "baz"
+		framework := Framework{BaseName: "cluster-upgrade"}
+		var webserver *WebserverTest
+
+		BeforeEach(func() {
+			framework.beforeEach()
+			webserver = NewWebserverTest(framework.Client, framework.Namespace.Name, name)
+			webserver.CreateWebserverPod()
+			svc := webserver.BuildServiceSpec()
+			svc.Spec.Type = api.ServiceTypeLoadBalancer
+			webserver.CreateService(svc)
+		})
+
+		AfterEach(func() {
+			framework.afterEach()
+			webserver.Cleanup()
+		})
+
+		It("by kube-push of master should maintain responsive services", func() {
+			testClusterUpgrade(framework, name, func() {
+				runUpgradeScript("hack/e2e-internal/e2e-push.sh", "-m")
+			})
+		})
+
+		It("by gce-upgrade of master should maintain responsive services", func() {
+			if !providerIs("gce") {
+				By(fmt.Sprintf("Skippingt test, which is not implemented for %s", testContext.Provider))
+				return
+			}
+			testClusterUpgrade(framework, name, func() {
+				runUpgradeScript("hack/e2e-internal/e2e-upgrade.sh", "-M", "-l")
+			})
+		})
+	})
+})
+
+func testClusterUpgrade(framework Framework, svcName string, upgrade func()) {
+	result, err := waitForLoadBalancerIngress(framework.Client, svcName, framework.Namespace.Name)
+	Expect(err).NotTo(HaveOccurred())
+	if len(result.Status.LoadBalancer.Ingress) != 1 {
+		Failf("got unexpected number (%v) of ingress points for externally load balanced service: %v", result.Status.LoadBalancer.Ingress, result)
+	}
+	ingress := result.Status.LoadBalancer.Ingress[0]
+	By("Waiting for pod to become reachable")
+	testLoadBalancerReachable(ingress, 80)
+
+	By("Upgrading cluster")
+	httpClient := http.Client{Timeout: 2 * time.Second}
+	retries := 3
```

Quinton has been advising us to switch from numbered retires to timed retries using `wait.Poll(...)`. We've got standard single call timeouts that you can use ([example](https://github.com/GoogleCloudPlatform/kubernetes/blob/bdeb4f31a8277323ee17cb54c769ae1fe161d8c1/test/e2e/util.go#L1088)), though maybe you want it to fail faster if you're trying to test availability with a more strict definition of "available."